osutil/disks/gpt.go

```diff
@@ -0,0 +1,134 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package disks
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"os"
+)
+
+type GPTLBA uint64
+type GPTGUID [16]byte
+
+type GPTHeader struct {
+	Signature	[8]byte
+	Revision	uint32
+	HeaderSize	uint32
+	HeaderCRC	uint32
+	Reserved	uint32
+	CurrentLBA	GPTLBA
+	AlternateLBA	GPTLBA
+	FirstUsableLBA	GPTLBA
+	LastUsableLBA	GPTLBA
+	DiskGUID	[16]byte
+	EntriesLBA	GPTLBA
+	NEntries	uint32
+	EntrySize	uint32
+	EntriesCRC	uint32
+}
+
+func LoadGPTHeader(devfd *os.File) (GPTHeader, error) {
+	var header GPTHeader
+	rawHeader := make([]byte, 512)
+	read, err := devfd.Read(rawHeader)
+	if err != nil {
+		return header, err
+	}
+
+	if read < binary.Size(header) {
+		return header, fmt.Errorf("Not enough data was read")
+	}
+
+	buf := bytes.NewReader(rawHeader[:read])
+	binary.Read(buf, binary.LittleEndian, &header)
+
+	if !bytes.Equal(header.Signature[:], []byte("EFI PART")) {
+		return header, fmt.Errorf("GPT Header does not start with magic string")
```

Comment:
from this point below, all the error conditions should return and empty GPTHeader{} here we may be returning an structure which is valid in terms of a Go structure and has data in the fields and such, but logically is corrupted because it doesn't have the right signature, crc doesn't match etc.

Reply:
Done.